### ISSUE TITLE
azure-cli create and upload ssh keys?

### ISSUE THREAD

**Issue by xied75** [Azure/azure-xplat-cli#2325]
Dear all,

I tried to find it but no success, can the cli create and upload local private keys to Azure for VM deployment use?

Best,

Dong

**Comment by squillace** [comment]
@xied75 Yes. See https://azure.microsoft.com/en-us/documentation/articles/virtual-machines-linux-use-ssh-key/#create-a-vm-with-your-public-key-file:

```
azure vm create \
-l "West US" -n testpemasm \
-P -t myCert.pem -e 22 \
testpemasm \
b39f27a8b8c64d52b05eac6a62ebad85__Ubuntu-14_04_3-LTS-amd64-server-20150908-en-us-30GB \
ops
info:    Executing command vm create
warn:    --vm-size has not been specified. Defaulting to "Small".
+ Looking up image b39f27a8b8c64d52b05eac6a62ebad85__Ubuntu-14_04_3-LTS-amd64-server-20150908-en-us-30GB
+ Looking up cloud service
info:    cloud service testpemasm not found.
+ Creating cloud service
+ Retrieving storage accounts
+ Configuring certificate
+ Creating VM
info:    vm create command OK
```

**Comment by squillace** [comment]
@xied75 if that's what you mean by "create and upload". of course, you must create your .pem files first. 

You can have the cli generate the keys for you:

`help:      -G, --generate-ssh-keys                                 Auto generate SSH keys, will be ignored if --ssh-cert is specified.`

**Comment by xied75** [comment]
@squillace no that's not what I'm asking, I didn't ask to upload a key while creating a vm, I need to upload the key as a single op.

**Comment by squillace** [comment]
So, @xied75, you have a key, and an existing ASM vm, and you want to upload that key to use it from a client machine?

**Comment by squillace** [comment]
if so, you need https://github.com/Azure/azure-linux-extensions/tree/master/VMAccess#211-classic. There is no direct call for this in the asm cli other than that one.

**Comment by xied75** [comment]
@squillace the whole point of raising issues here is to bring the discussion and eventually implement it or reject with a reason. I know it can't be done currently.

Ssh keys in other clouds are standalone entity, it does not 'belong' to anything. Therefore the creation shall be a single step too. During vm deployment one simply refer to the name of the key. So if I deploy 100 vms they can all refer to the same key created beforehand.

**Comment by squillace** [comment]
Forgive me @xied75, but I'm still not clear what you're looking for. SSH keys on Azure have nothing to do with Azure; they're an SSH connection thing. You could use SSH with usernames and passwords if it were secure enough; but because Linux VMs are best secured with ssh-rsa or .pem public key files, you simply pass the same .pub or .pem file to all 100 VMs in Azure when you create them as you would any other place, in the cloud or locally. All the CLI does is what any other Linux creation system would do, which is to install the SSH information in `authorized_users`. 

There are two issues I can imagine that you might be bringing up. 

1. You want a single call in the ASM mode of the CLI to install a new user and public key file into `authorized_users`. There is this call in the ARM mode (`azure vm reset-access`); in the ASM mode, there is also one call, but it uses a bit of json to work: `azure vm extension set <vm-name> \
VMAccessForLinux Microsoft.OSTCExtensions <version> \
--private-config-path protected.json`
2. You are actually looking for the correct way to have one file called `myPublicKeyFile.pub` (or .pem) and create any number of VMs using that key. If this is the case, then the link above to use `azure vm create` in ASM mode and passing your public key file is the way to do that. You can use the same public key file for every single VM you create if you want. This has absolutely nothing to do with Azure at all. 

I am happy to raise the issue about anything here! However, I want to be clear what, precisely, it is that you are arguing for -- you can create any number of VMs with a .pub or .pem key in ASM from the CLI, and you can use one call to insert new key files and users in ASM from the CLI. 

The only thing that I can see is true is that there is no `azure vm reset-access` command in the ASM mode of the CLI. That is undeniably true. But because there's a documented single call to do the same thing, I'm not sure how important **I** think that is. However, I'm not the engineering team; you can make your argument here for sure and they'll be happy to review.

**Comment by xied75** [comment]
http://docs.aws.amazon.com/cli/latest/reference/ec2/create-key-pair.html
http://docs.aws.amazon.com/AWSEC2/latest/CommandLineReference/ApiReference-cmd-CreateKeyPair.html
https://cloudstack.apache.org/api/apidocs-4.4/user/registerSSHKeyPair.html
http://shankerbalan.net/blog/ssh-keys-on-cloudstack-guest-instances-using-cloud-init/
http://docs.openstack.org/cli-reference/content/novaclient_commands.html#novaclient_subcommand_keypair-add

**Comment by squillace** [comment]
@xied75, we **HAVE** "create a key pair for me" in the CLI. You use the `help: -G, --generate-ssh-keys Auto generate SSH keys, will be ignored if --ssh-cert is specified` switch on the `azure vm create` command, and the key pair is generated for you on your machine AND added to the vm that you are creating. You can then reuse that keypair, or you can continue generating new keypairs for each machine you create. It's up to you.

You seem to be requesting that we have a specific switch to create a keypair in the CLI that does nothing else at all (this is what I read those documents saying: that they have a call to create a keypair and do nothing else). Is that what you would like?

**Comment by squillace** [comment]
@xied75 I think I'll close this, but if I didn't get what you are looking for, reopen it! @yugangw-msft do you want to close for now?

**Comment by xied75** [comment]
@squillace I'm pretty sure I haven't got what I'm looking for. :) You are correct in your 2nd paragraph that I want a separate command that does nothing more than just import the key. It seems to me that you treat key management as second class citizen compared with creating a vm, so you insist to tie these two together, I guess tons of enterprises would not like that, otherwise why would azure create the "key vault" product?

**Comment by jonathan-kosgei** [comment]
To create the key:
```
openssl req -x509 \
  -key ~/.ssh/id_rsa \
  -nodes \
  -days 365 -newkey rsa:2048 \
  -out ~/.ssh/azure.pem \
  -subj '/CN=www.company.com/O=Company./C=KE'

openssl x509 \
  -outform der \
  -in ~/.ssh/azure.pem \
  -out ~/.ssh/azure.pfx

fingerprint=$(openssl x509 -fingerprint -inform der -in ~/.ssh/azure.pfx | grep "SHA1 Fingerprint")
fingerprint="${fingerprint#*=}"
fingerprint="${fingerprint//:/}"
echo $fingerprint
```

The last 4 lines generate a fingerprint you can use for terraform. To upload the key after creating the vm run:` azure service cert create {hosted_service_provider_name} {path to azure.pfx}`

**Comment by xied75** [comment]
@jonathan-kosgei Thanks for posting! I suppose this is what we asked half year ago? That we can upload keys/certs without creating a VM?

Interestingly the command you show is in ASM mode, do we have something equivalent in ARM?

Best regards,

**Comment by amarzavery** [comment]
@huangpf @sauryadas - Any updates on this?

**Comment by jonathan-kosgei** [comment]
@xied75 Hi, not really sure about that.
